Add playwright e2e tests

nicodh opened the pull request:
Start having e2e tests with playwright based on the browser view

see [README](https://github.com/deltachat/deltachat-desktop/tree/ndh/playwright-e2e-tests/packages/e2e-tests#readme)

Simon-Laux commented on packages/e2e-tests/README.md:
```diff
@@ -0,0 +1,29 @@
+
+## Usage
+
+`pnpm e2e --project <chromium | firefox | Chrome>` for headless usage
```
```suggestion
`pnpm -w e2e --project <chromium | firefox | Chrome>` for headless usage
```

-w to run the version on project-root to also include the build

Simon-Laux commented on packages/e2e-tests/README.md:
```diff
@@ -0,0 +1,29 @@
+# End to end tests
+
+## Install playwright
+
+cd into packages/e2e-tests
+
+and run `pnpm create playwright`
+
+in the upcoming dialog
+
+selects "tests" as target for tests and skip the creation of playwright.config.ts
+
+after the install remove the created tests/examples and packages/e2e-tests/tests/example.spec.ts
```
I doubt that this is a good solution, won't `npx playwright install --with-deps` suffice?

Simon-Laux commented on packages/target-electron/package.json:
```diff
@@ -40,7 +40,7 @@
     "dev": "pnpm build && pnpm start --",
     "dev-inspect": "pnpm build && pnpm start --inspect --",
-    "start": "NODE_OPTIONS=--enable-source-maps electron . --devmode --disable-http-cache --translation-watch",
+    "start": "NODE_OPTIONS=--enable-source-maps electron . --disable-http-cache --translation-watch",
```
why remove `--devmode` here?

nicodh replied:
Just an uninteded leftover from a local test

WofWca commented on packages/e2e-tests/.github/workflows/playwright.yml:
```diff
@@ -0,0 +1,27 @@
+name: Playwright Tests
```
Does this file do anything? I thought `.github` is supposed to be at the top level of the repo.

nicodh replied:
It's an auto generated github action that we will integrate in another step later

WofWca commented:
Maybe it makes sense to also update other READMEs, e.g. ["How to do end to end testing"](https://github.com/deltachat/deltachat-desktop/blob/21002182831abf909726771d82e65f9cec0f649d/README.md#L38), and the document that it's referencing.

nicodh commented:
> Maybe it makes sense to also update other READMEs, e.g. ["How to do end to end testing"](https://github.com/deltachat/deltachat-desktop/blob/21002182831abf909726771d82e65f9cec0f649d/README.md#L38), and the document that it's referencing.

Good point! I updated it

nicodh commented:
> tests don't start when I also have the browser version open (they should use different default port to fix that I guess)

I added a line in the docs not to start the browser since that happens in the tests

WofWca commented:
About READMEs again.

- [ ] There is this section, maybe it needs to be updated as well:

https://github.com/deltachat/deltachat-desktop/blob/4d8f77c73e76bdcc157f81d6a68c664aaa61d17a/docs/DEVELOPMENT.md#L104-L110

- [ ] And here is a section about env variables:

https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/docs/CLI_FLAGS.md#L22-L29

Also, what about the old tests? This MR removes some of the old testing code, but not all of it (search for `testcafe` and `TEST_DIR`. Is it for later?

nicodh commented:
> Also, what about the old tests? This MR removes some of the old testing code, but not all of it (search for `testcafe` and `TEST_DIR`. Is it for later?

Good findings! I updated the relevant docs and removed unused test stuff

WofWca commented on packages/e2e-tests/env.example:
```diff
@@ -0,0 +1 @@
+WEB_PASSWORD=your-password
```
Is this file needed here?

WofWca reviewed (changes requested):
Tests run for me with `pnpm -w e2e --project chromium`, and with the Playwright extension, in UI mode as well!
Though some warnings are printed and they looks scary, e.g. `src/context.rs:465: can not start io on a context that is not configured`.

The changes to `packages/frontend` (i.e. runtime, production code) are minor ind should have no practical effect, i.e. probably safe to merge and release.

The changes to `packages/target-electron` also seem moderate, except maybe for the `dotenv` dependency that I'm not sure is tree-shaken.

I haven't looked at the actual test code, but it appears to be working! And it doesn't affect production.

Apart from a weird `aria-label` change, I think it's safe to merge for the upcoming production release.

nicodh commented on packages/frontend/src/components/Dialog/HeaderButton.tsx:
```diff
@@ -13,6 +13,7 @@ type Props = ButtonHTMLAttributes<HTMLButtonElement> & {
   iconSize: number
   rotation?: number
   'aria-label': string
+  dataTestid?: string
```
Hmm, how is it included in ButtonHTMLAttributes which comes from react?
If I remove it in the type defintion typescript complains about an unknown prop testId...?

Simon-Laux replied:
you can probably extend that ButtonHTMLAttributes with the additional attribute in global.d.ts, similar how we do it with our `window` extensions.

Simon-Laux replied:
but we can also do that in a follow up pr

WofWca replied:
Make sure to use kebab-case, i.e.

```typescript
<Button
  data-testid='test'
>
```

for me it doesn't complain this way.

nicodh replied:
It works anywhere where the {...props} attribute is passed down to the last rendered element, which is the case for Buttons. So you were right and I removed the testId props for buttons

nicodh commented on packages/target-browser/.env.example:
```diff
@@ -0,0 +1,7 @@
+WEB_PASSWORD=
+
+# optional
+# with NODE_ENV=test the login is skipped
```
I don't think its worth. The browser is only for development anyway

nicodh replied:
I changed the comment from the frist proposal

WofWca replied:
> The browser is only for development anyway

IDK IDK. From @Simon-Laux 's comments here and there I'm under the impression that it's something that is maybe gonna be in production some day.

Simon-Laux replied:
if someone makes it production they can still make it more secure then. 🤔

nicodh commented on packages/target-electron/src/application-config.ts:
```diff
@@ -1,11 +1,17 @@
 import applicationConfig from 'application-config'
+if (process.env.NODE_ENV !== 'production') {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const { config } = require('dotenv')
```
its in devDependencies anyway, so it won't be installed in production

Simon-Laux replied:
> its in devDependencies anyway, so it won't be installed in production

that's not how it works here. esbuild bundles it. but I wonder more why require is even allowed here because this is an esm module, I thought there was some workaround to required to get the original `require`

so there is still the tree shake questions remaining, is this still in the bundle in production builds? afaik only the frontend has `NODE_ENV=production` with `pnpm run build4production` or have I missed something.

I would just make this dependent on `--devmode` cli option unless we really have a good workflow that excludes it in production.

Simon-Laux replied:
we can convert this into an issue and merge this already for the test release

nicodh replied:
I just tried await import('dotenv') instead and it seems to work: found that here: https://dmitripavlutin.com/ecmascript-modules-dynamic-import/

I will test again in production mode

nicodh replied:
It works as expected. Even if I uninstall dotenv I can build for production. Added a try catch to make sure it won't influence prod build

nicodh commented on packages/e2e-tests/package.json:
```diff
@@ -0,0 +1,14 @@
+  "name": "@deltachat-desktop/e2e-tests",
+  "type": "module",
+  "license": "GPL-3.0-or-later",
+  "version": "1.46.2-monorepo",
```
I updated the version. The license was in there before. Maybe discuss it with Simon...

nicodh commented on docs/E2E-TESTING.md:
```diff
@@ -1,7 +1,35 @@
-`.test-*` - to mark special elements so the test can find them.
+cd into packages/e2e-tests
+
+and run `npx playwright install --with-deps`
```
It will only load playwright and execute it. No idea if it has impact, I don't think so
